File: Leetcode/77.cpp
/**
 * 77. 组合 深搜 + 剪枝
 * link：https://leetcode-cn.com/problems/combinations/
 */

class Solution
{
    int numbers;
    int target;
    vector<vector<int>> result;
    vector<int> path;

public:
    vector<vector<int>> combine(int n, int k)
    {
        numbers = n;
        target = k;

        dfs(1);
        return result;
    }

    void dfs(int begin)
    {
        int n = path.size();
        if (n == target)
        {
            result.push_back(path);
            return;
        }

        for (int i = begin; i <= numbers - target + n + 1; i++)
        {
            path.push_back(i);
            dfs(i + 1);
            path.pop_back();
        }
    }
};

File: Leetcode/77.py
# 77. 组合 深搜 + 剪枝
# link：https://leetcode-cn.com/problems/combinations/

class Solution:
    def combine(self, n: int, k: int) -> List[List[int]]:
        if k == 0:
            return []
        elif k == 1:
            return [[i]for i in range(1, n+1)]
        elif k == n:
            return [[i for i in range(1, n+1)]]

        path = []
        result = []

        def dfs(begin: int):
            if len(path) == k:
                result.append(path.copy())
                return
            for i in range(begin, n - k + len(path)+2):
                path.append(i)
                dfs(i+1)
                path.pop(len(path)-1)

        dfs(1)
        return result
